routers/api/v1/repo/repo.go

```diff
@@ -697,6 +698,17 @@ func updateBasicProperties(ctx *context.APIContext, opts api.EditRepoOption) err
 		repo.IsTemplate = *opts.Template
 	}
 
+	// Default branch only updated if changed and exist
+	if opts.DefaultBranch != nil && repo.DefaultBranch != *opts.DefaultBranch && ctx.Repo.GitRepo.IsBranchExist(*opts.DefaultBranch) {
+		if err := ctx.Repo.GitRepo.SetDefaultBranch(*opts.DefaultBranch); err != nil {
+			if !git.IsErrUnsupportedVersion(err) {
```

Comment:
What should happen if it _is_ an unsupported version?

Reply:
Currently I implemented in same way as for UI; git HEAD ref is not updated but the default branch is set in the database. What is the end results? No idea, anyone tested?

Reply:
https://github.com/go-gitea/gitea/blob/722a7c902dd39bd3d4328345ca969220640774d7/modules/git/repo_branch.go#L68-L72

I don't think it can return an unsupported error... Symbolic-ref is present in 1.7.2

Reply:
I think what will happen is that since HEAD is not changed, when you clone you will still get "master" checked-out, but in the Gitea UI the new default branch is used. So I think it makes sense to allow changing default branch also for earlier git.